### ISSUE TITLE
argument `training` hard set to `True` in call to F.batch_norm

### ISSUE THREAD

**Issue by simonguiroy** [AntreasAntoniou/HowToTrainYourMAMLPytorch#3]
It should be `training=training` instead of `training=True`, since here even at inference time, the BN function will use data statistics instead of moving statistics.

https://github.com/AntreasAntoniou/HowToTrainYourMAMLPytorch/blob/09dbacc7415bb0659f84f40d5c264179d80a9bd6/meta_neural_network_architectures.py#L269

**Comment by AntreasAntoniou** [comment]
In Pytorch, using training=True, does not use the data stats, but actually updates the moving stats using the data stats and uses that to apply BN.

Now there is another question to answer. Isn't it cheating for me to store info that came from validation data? Yes, it is. But it is done in a way that prevents me from cheating. You actually want to use information available from a task, to do well in that task. Using the stored running stats alone would not utilize the new information you have. So I instead allow evaluation data stats to be used to update the running means that are used to compute the target task's statistics. However, once a task has been completed (i.e. inner loop steps + inference on target task), I reset the running stats to be the last available training-sourced stats (i.e. the stats at the last training iter, before evaluation began). Using this little trick, I allow my model to use data distribution information from a new task at inference time without 'contaminating' the saved stats with validation stats.

**Comment by namsan96** [comment]
Hi,
I also thought the training argument should not be set to always True.
If the referred line line is only used to save data stats, then where is the part that actually use it?

**Comment by jfb54** [comment]
> In Pytorch, using training=True, does not use the data stats, but actually updates the moving stats using the data stats and uses those updated moving stats to apply BN.
> 
This is not correct. When calling F.batch_norm with training=True, the batch statistics will be used to normalize, not the running statistics. The running statistics will only be used to normalize when training=False. You are correct in saying that setting training=True will update the running stats, but it won't apply them unless training=False. Thus, the code as written will not utilize Per-Step Batch Normalization Running Statistics (BNRS) as described in the paper.
